backend: format backend logs to use zlog

knrt10 opened the pull request:
Earlier we were not using zlog everywhere due to which some logs were not correctly parsed by log platforms. They were all parsed as errors. This fixes that and adds correct format of all logs in headlamp. They are now shown as json which can be parsed by any log platform.

This also adds a wrapper function for zlog which increases code readability and cleanliness.

Fixes: #1645

yolossn commented on backend/pkg/helm/handler.go:
```diff
@@ -39,7 +39,10 @@ func NewActionConfig(clientConfig clientcmd.ClientConfig, namespace string) (*ac
 	}
 	logger := func(format string, a ...interface{}) {
-		log.Info().Str("namespace", namespace).Msg(format + "\n" + fmt.Sprintf("%v", a))
+		zlog.Info().
```
I think instead of printing the action manually you can create a helper function which internally uses the golang runtime packages [Caller](https://pkg.go.dev/runtime#Caller) function to print the source of the log. That can be helpful in identifying the source of the log very easily.

knrt10 replied:
Let me look into it

yolossn reviewed (changes requested):
I think it will be a better idea to create a helper interface or function to print the logs. It will make the code more cleaner and will make it easier for us to migrate to another logging package incase we want to migrate in the future. Also gives us the advantage of using the runtime package to print the source of the log which will make it easier to identify from where the log was created. WDYT?

knrt10 commented:
Thanks for the suggestion @yolossn. I have updated the code as PTAL. Thanks

yolossn commented on backend/cmd/headlamp.go:
```diff
@@ -361,19 +368,15 @@ func createHeadlampHandler(config *HeadlampConfig) http.Handler {
-				Str("action", "externalproxy").
-				Str("proxyURL", proxyURL).
-				Msg("proxy URL is empty")
+			str := map[string]string{"proxyURL": proxyURL}
+			utils.Log(utils.LevelError, str, errors.New("proxy URL is empty"), "proxy URL is empty")
```
```suggestion
			utils.Log(utils.LevelError, map[string]string{"proxyURL": proxyURL}, errors.New("proxy URL is empty"), "proxy URL is empty")
```

yolossn commented on backend/cmd/headlamp.go:
```diff
@@ -361,19 +368,15 @@ func createHeadlampHandler(config *HeadlampConfig) http.Handler {
-				Str("action", "externalproxy").
-				Str("proxyURL", proxyURL).
-				Msg("The provided proxy URL is invalid")
+			str := map[string]string{"proxyURL": proxyURL}
+			utils.Log(utils.LevelError, str, err, "The provided proxy URL is invalid")
```
```suggestion
			utils.Log(utils.LevelError, map[string]string{"proxyURL": proxyURL}, err, "The provided proxy URL is invalid")
```

yolossn commented on backend/cmd/headlamp.go:
```diff
@@ -463,20 +471,26 @@ func createHeadlampHandler(config *HeadlampConfig) http.Handler {
 		kContext, err := config.kubeConfigStore.GetContext(cluster)
 		if err != nil {
-			log.Printf("Error: failed to get context: %s", err)
+			str := map[string]string{"cluster": cluster}
+			utils.Log(utils.LevelError, str, err, "Error: failed to get context")
+
```
```suggestion
			utils.Log(utils.LevelError, map[string]string{"cluster": cluster}, err, "Error: failed to get context")

```

yolossn commented on backend/cmd/headlamp.go:
```diff
@@ -463,20 +471,26 @@ func createHeadlampHandler(config *HeadlampConfig) http.Handler {
 		oidcAuthConfig, err := kContext.OidcConfig()
 		if err != nil {
-			log.Printf("Error getting %s cluster oidc config %s", cluster, err)
+			str := map[string]string{"cluster": cluster}
```
similar to other comments no need to create the map as a separate variable since it is used only as an argument to the `utils.Log` function.

yolossn commented on backend/cmd/headlamp.go:
```diff
@@ -675,17 +709,23 @@ func refreshAndCacheNewToken(oidcAuthConfig *kubeconfig.OidcConfig,
 
 	rToken, ok := refreshToken.(string)
 	if !ok {
+		utils.Log(utils.LevelError, nil, err, "Error: failed to cast refresh token")
```
```suggestion
		utils.Log(utils.LevelError, nil, err, "failed to cast refresh token")
```

since it is passed with Error as severity I think it is not necessary to add Error to the message.

yolossn commented on backend/cmd/headlamp.go:
```diff
@@ -675,17 +709,23 @@ func refreshAndCacheNewToken(oidcAuthConfig *kubeconfig.OidcConfig,
 	// get refresh token from cache
 	refreshToken, err := cache.Get(context.Background(), fmt.Sprintf("oidc-token-%s", token))
 	if err != nil || refreshToken == "" {
+		utils.Log(utils.LevelError, nil, err, "Error: failed to get refresh token")
```
```suggestion
		utils.Log(utils.LevelError, nil, err, "failed to get refresh token")
```

yolossn commented on backend/cmd/headlamp.go:
```diff
@@ -703,19 +743,25 @@ func refreshAndCacheNewToken(oidcAuthConfig *kubeconfig.OidcConfig,
 
 	tk, err := ts.Token()
 	if err != nil {
+		utils.Log(utils.LevelError, nil, err, "Error: failed to get new token")
```
```suggestion
		utils.Log(utils.LevelError, nil, err, "failed to get new token")
```

yolossn commented on backend/cmd/headlamp.go:
```diff
@@ -703,19 +743,25 @@ func refreshAndCacheNewToken(oidcAuthConfig *kubeconfig.OidcConfig,
 	if ok {
 		// update cache
 		if err := cache.Set(context.Background(), fmt.Sprintf("oidc-token-%s", idToken), tk.RefreshToken); err != nil {
+			utils.Log(utils.LevelError, nil, err, "Error: failed to cache refresh token")
```
```suggestion
			utils.Log(utils.LevelError, nil, err, "failed to cache refresh token")
```
Make sure to remove the extra "Error" every where the log level is set to Error.

yolossn commented on backend/pkg/utils/utils.go:
```diff
@@ -10,3 +17,52 @@ func Contains[T comparable](elems []T, v T) bool {
+		// Check if err is a error slice ([]error)
+		if errs, ok := err.([]error); ok {
+			event.Errs("error", errs).Msg(msg)
+		} else {
+			// Handle other error types
+			event.Err(err.(error)).Msg(msg)
+		}
```
The code will reach this state even when a integer is passed as err. Either handle the state properly here or use custom type supported by Golang instead of interface in the function argument.

yolossn reviewed (changes requested):
There are few nits apart from that it looks good to me.

illume commented on backend/pkg/kubeconfig/watcher.go:
```diff
@@ -95,7 +101,8 @@ func addFilesToWatcher(watcher *fsnotify.Watcher, paths []string) {
 		if err != nil {
-			log.Printf("Couldn't add %s to watcher: %v", path, err)
+			utils.Log(utils.LevelError, map[string]string{"path": path},
+				err, "Error adding path to watcher")
```
I guess there needs to be some consistency in the logs with regards to "Error" being mentioned in the message. 

I expect "Error" should not be used, since it's removed from some of the messages?

illume commented on backend/pkg/helm/handler.go:
```diff
@@ -39,7 +39,8 @@ func NewActionConfig(clientConfig clientcmd.ClientConfig, namespace string) (*ac
 	logger := func(format string, a ...interface{}) {
-		log.Info().Str("namespace", namespace).Msg(format + "\n" + fmt.Sprintf("%v", a))
+		utils.Log(utils.LevelInfo, map[string]string{"namespace": namespace},
+			nil, format+"\n"+fmt.Sprintf("%v", a))
```
Should this Sprintf still be there? I think a is passed into the logger already.

knrt10 replied:
I think yes, since the last part of Log function is for message so `format+"\n"+fmt.Sprintf("%v", a)` should be kept as it is.

illume reviewed (changes requested):
Nice one.

I left a few questions/notes.

illume commented on backend/pkg/utils/utils.go:
```diff
@@ -10,3 +17,58 @@ func Contains[T comparable](elems []T, v T) bool {
+// Log is a wrapper function for logging. It uses zlog package and logs to stdout.
+// It logs the message, source file and line number.
+// It logs the message at the level specified.
+func Log(level uint, str map[string]string, err interface{}, msg string) {
```
Can you please add test(s) for this function?

illume commented:
Leaving this topic @joaquimrocha and I discussed about this being a breaking change... It might potentially break someone somewhere, but IMHO I think it's not a wide spread break. We should mention the change to the log format in the release notes.

illume commented:
Just checking... have you confirmed the logs are compatible with this link Julian left? https://learn.microsoft.com/en-us/azure/developer/java/sdk/logging-overview#default-logger-for-temporary-debugging

I wonder specifically if these are compatible? https://pkg.go.dev/github.com/rs/zerolog#pkg-variables
- field names ( LevelFieldName, MessageFieldName, TimestampFieldName, CallerFieldName )
- time stamp (TimeFieldFormat)

knrt10 commented on backend/pkg/helm/release.go:
```diff
@@ -8,6 +8,7 @@ import (
 	"github.com/gorilla/schema"
+	"github.com/headlamp-k8s/headlamp/backend/pkg/utils"
 
 	"github.com/rs/zerolog"
```
There are some instances in this file were zerolog is used differently than passing as a logging function, that's why it's imported here. Please check https://github.com/headlamp-k8s/headlamp/blob/main/backend/pkg/helm/release.go#L683

knrt10 commented:
> Just checking... have you confirmed the logs are compatible with this link Julian left? https://learn.microsoft.com/en-us/azure/developer/java/sdk/logging-overview#default-logger-for-temporary-debugging

yes, I pinged him regarding this, we have implemented zlog everywhere which gives JSON data for errors, warning, info and such. It is standard with all the tools since JSON is easily parse able everywhere. Also, if user's usecase is other than JSON logs we have also implemented a wrapper function Log [backend: format backend logs to use zlog by knrt10 · Pull Request #1723 · headlamp-k8s/headlamp (github.com)](https://github.com/headlamp-k8s/headlamp/pull/1723/files#diff-a205007fb04a23a21624acc3bde58fe4d346f7a52fc7ceccf709f3b1dfd2dc6cR33) which can be easily modified as per their needs.

illume reviewed (approved):
🎉 🎈
